### PR TITLE
[SPARK-17703][SQL] Add unnamed version of addReferenceObj for minor objects.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

There are many minor objects in references, which are extracted to the generated class field, e.g. `errMsg` in `GetExternalRowField` or `ValidateExternalType`, but number of fields in class is limited so we should reduce the number.
This pr adds unnamed version of `addReferenceObj` for these minor objects not to store the object into field but refer it from the `references` field at the time of use.
## How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -85,6 +85,27 @@ class CodegenContext {
...
+  def addReferenceObj(obj: Any): String = {
+    addReferenceObj(obj, null)
+  }
+
+  /**
+   * Add an object to `references`.
+   *
+   * Returns the code to access it.
+   */
+  def addReferenceObj(obj: Any, className: String): String = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: where is this used?

Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: This is used by another `addReferenceObj(obj: Any)` method defined above.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Yea but looks like it is always null. Why not just remove this and move the body into the previous function?

Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I thought that there would be the case we want to specify the class name for the object, the same as the original (named) one.
Don't we need this?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Maybe add it when we need it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -85,6 +85,27 @@ class CodegenContext {
   val references: mutable.ArrayBuffer[Any] = new mutable.ArrayBuffer[Any]()
 
   /**
+   * Add an object to `references`.
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: also let's expand the comment to discuss the behavior, and why it is needed (basically your pr description)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 1 above)
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I see, I'll remove and move the body.
Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I see, I'll expand the comment.
Thanks.

**Comment by rxin** [comment]
BTW did you run into an issue? My worry is that this does make the code slightly less clear.

**Comment by ueshin** [comment]
@rxin Yes, in my case, I used `RowEncoder` with a schema having a lot of fields (>250) and processed some logic. `RowEncoder` uses `GetExternalRowField` and `ValidateExternalType` and they add 2 instance variables per field, serialize. I think this pr will make things better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 3]
```diff
@@ -906,7 +906,7 @@ case class AssertNotNull(child: Expression, walkedTypePath: Seq[String])
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val childGen = child.genCode(ctx)
 
-    val errMsgField = ctx.addReferenceObj("errMsg", errMsg)
+    val errMsgField = ctx.addReferenceObj(errMsg)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: thanks - then i'd actually add a comment in the caller indicating it is important to call the variant that doesn't create a local field.

Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: I see, I'll add comments.

**Comment by rxin** [comment]
Thanks - merging in master.

**Comment by davies** [comment]
@rxin  Hit the same issue here today (10k errMsg in generated UnsafeProjection), should be back port this into 2.0 branch?

**Comment by rxin** [comment]
Sure sounds good.

**Comment by ueshin** [comment]
@davies @rxin Should I send a PR to backport this to `branch-2.0` if you don't have time?

**Comment by rxin** [comment]
Please do! I was hoping you'd do it...!

**Comment by ueshin** [comment]
Oh, I see, wait for a while.

**Comment by ueshin** [comment]
I sent a new PR #15808. Thanks.

**Pull request closed without merging**